### ISSUE TITLE
A problem about the DDPG

### ISSUE THREAD

**Issue by geweihgg** [pemami4911/deep-rl#15]
Hi, I want to implement the DDPG algorithm, and before that, i've read your code. It's very useful. And I still have some little questions about the code.
1.As in the DDPG.py, line 61 to 66:
        # This gradient will be provided by the critic network
        self.action_gradient = tf.placeholder(tf.float32, [None, self.a_dim])
        # Combine the gradients here
        self.actor_gradients = tf.gradients(
                self.scaled_out, self.network_params, -self.action_gradient)
My problem is: The tf.gradients(y,x) sums up the dy/dx for each y in ys. And in the paper, the equation of dJ/d(theta) divides N, and I wonder that whether I should write the code like this:
        self.actor_gradients = tf.div(tf.gradients(
                self.scaled_out, self.network_params, -self.action_gradient),N)
Looking forward to your reply.
Thank you very much.
My contact way: email: 249806873@qq.com

**Comment by pemami4911** [comment]
Hey- no need to divide by N. The gradients are already averaged over the batch size for you. Check out this discussion - https://github.com/pemami4911/deep-rl/issues/3

**Comment by geweihgg** [comment]
Thank you very much!
I still don't understand that "The gradients are already averaged over the batch size", and I wrote something like this:
`import tensorflow as tf
sess=tf.InteractiveSession()
x=tf.placeholder(dtype=tf.float32,shape=[2,2])
w=tf.constant(value=[2,3,5,6],dtype=tf.float32,shape=[2,2])
x_input=[[1,5],[3,4]]
y=tf.matmul(x,w)
grad_y=tf.gradients(y,w)
sess.run(tf.global_variables_initializer())
print len(grad_y)
print grad_y[0].eval(feed_dict={x:x_input})`
Then the output is like this:
`1
[[4. 4.]
[9. 9.]]`
It seems that the tf.gradients() doesn't divide the gradients by N, thought it sums up over the batch.

Looking forward to your reply.
Thank you again.

**Comment by geweihgg** [comment]
Thank you very much!
I still don't understand that "The gradients are already averaged over the batch size", and I wrote something like this:
`import tensorflow as tf 
sess=tf.InteractiveSession() 
x=tf.placeholder(dtype=tf.float32,shape=[2,2]) 
w=tf.constant(value=[2,3,5,6],dtype=tf.float32,shape=[2,2]) 
x_input=[[1,5],[3,4]] 
y=tf.matmul(x,w) 
grad_y=tf.gradients(y,w) 
sess.run(tf.global_variables_initializer()) 
print len(grad_y) 
print grad_y[0].eval(feed_dict={x:x_input})`
Then the output is like this:
`1 
[[4. 4.] 
[9. 9.]]`
It seems that the tf.gradients() doesn't divide the gradients by N, thought it sums up over the batch.

Looking forward to your reply.
Thank you again.

**Comment by geweihgg** [comment]
Thank you very much!
I still don't understand that "The gradients are already averaged over the batch size", and I wrote something like this:

import tensorflow as tf 
sess=tf.InteractiveSession() 
x=tf.placeholder(dtype=tf.float32,shape=[2,2]) 
w=tf.constant(value=[2,3,5,6],dtype=tf.float32,shape=[2,2]) 
x_input=[[1,5],[3,4]] 
y=tf.matmul(x,w) 
grad_y=tf.gradients(y,w) 
sess.run(tf.global_variables_initializer()) 
print len(grad_y) 
print grad_y[0].eval(feed_dict={x:x_input})

Then the output is like this:

1 
[[4. 4.] 
[9. 9.]]

It seems that the tf.gradients() doesn't divide the gradients by N, thought it sums up over the batch.

Looking forward to your reply.
Thank you again.

**Comment by pemami4911** [comment]
Deleted my previous comment to avoid confusion - it's been a while since I've looked at this sorry! #3 was in reference to `tf.gradients(self.out, self.action)`. In general, `tf.gradients` does sum over a batch, but when dyi/dxj = 0 when i ~= j (as is the case with `tf.gradients(self.out, self.action)`, you don't need to worry about dividing by the batch size since all terms except one for each i are 0. I have to look at `tf.gradients(self.scaled_out, self.network_params, -self.action_gradients)` more closely later today or tomorrow and get back to you

**Comment by geweihgg** [comment]
Hi~
I think I've figured out this problem.
I have misused the tf.gradients() function. A distinction must be made between "dQ/da" and "da/d(theta)".
In "dQ/da", we are computing the gradients wrt. the input, so there is no need to worry about the batch, cause tf.gradients() doesn't sum up over the batch.
But in "da/d(theta)", we computes the gradients wrt. the parameters, thus the tf.gradients() sums up over the batch. And we can solve this by dividing the results by N (i.e. the batch_size), or we can divide the learning rate by N.
What do you think about it?

**Comment by pemami4911** [comment]
I believe you are correct, I'll double check myself in a bit

**Comment by geweihgg** [comment]
Thank you again~
Keep in touch later~
